plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarRecordCursor.java

```
@@ -0,0 +1,734 @@
+                    currentRowValuesMap.put(columnHandle, bytesValueProvider(
+                            currentMessage.getMessageId().toString().getBytes(StandardCharsets.UTF_8)));
+                }
+                else if (PulsarInternalColumn.SEQUENCE_ID.getName().equals(columnHandle.getName())) {
+                    currentRowValuesMap.put(columnHandle, longValueProvider(currentMessage.getSequenceId()));
+                }
+                else if (PulsarInternalColumn.PRODUCER_NAME.getName().equals(columnHandle.getName())) {
+                    currentRowValuesMap.put(columnHandle,
+                            bytesValueProvider(currentMessage.getProducerName().getBytes(StandardCharsets.UTF_8)));
+                }
+                else if (PulsarInternalColumn.KEY.getName().equals(columnHandle.getName())) {
+                    String key = currentMessage.getKey().orElse(null);
+                    currentRowValuesMap.put(columnHandle, bytesValueProvider(key == null ? null :
+                            key.getBytes(StandardCharsets.UTF_8)));
+                }
+                else if (PulsarInternalColumn.PROPERTIES.getName().equals(columnHandle.getName())) {
+                    try {
+                        currentRowValuesMap.put(columnHandle, bytesValueProvider(
+                                new ObjectMapper().writeValueAsBytes(currentMessage.getProperties())));
+                    }
+                    catch (JsonProcessingException e) {
+                        throw new RuntimeException("Fail to decode Pulsar properties column", e);
+                    }
+                }
+                else {
+                    throw new IllegalArgumentException("unknown internal field " + columnHandle.getName());
+                }
+            }
+        }
+        try {
+            currentMessage.release();
+        }
+        catch (IllegalReferenceCountException e) {
```

please do not swallow exceptions